### PR TITLE
Calculating joint score and likelihood ratio from trace

### PR DESCRIPTION
**Pull request by johannbrehmer**
Work in progress -- doesn't work yet.

The main missing piece right now is that I don't know how to reconstruct the relation between the simulator parameters and the parameters of the distributions of the latent variables. This graph is apparently not saved in the trace...

### CODE REVIEW

**Comment by jan-matthis** [comment]
Is the PR still work in progress or ready for review?

**Comment by johannbrehmer** [comment]
Still in progress, I haven't had time to work out the trace replaying thing... I'll finish it this week!

Also, I'm not sure where the analyse_trace.py should ideally go. It serves a different purpose than the simulator codes. Should it still go in the benchmark module?

**Comment by jan-matthis** [comment]
> Also, I'm not sure where the analyse_trace.py should ideally go. It serves a different purpose than the simulator codes. Should it still go in the benchmark module?

I think the `augmented_data` method on the simulator class from your last commit is a neat solution. :)

**Comment by johannbrehmer** [comment]
Okay,

so I think the joint likelihood ratio is working. Thanks for teaching me Pyro, the ability to replay traces is really cool!

I can't get the score to come out right, though. I'm not sure what I'm doing wrong or whether there is some inherent problem with the computational graph of the log_probs of the Pyro distributions. I made a minimal working example in examples/wrong_score_mwe.py, maybe you can have a look and let me know what you think?

Thanks!
Johann

**Comment by jan-matthis** [comment]
I had a quick look: If you replace `x = pyro.sample("normal", normal)` by `x = pyro.sample("normal", normal).detach()` or `x = normal.sample()` results are correct. 

I'm not clear on why we'd need to detach `x` from the graph if we are using `pyro.sample()`..

**Comment by johannbrehmer** [comment]
That's wild. How on earth did you come up with that solution? Thanks a lot, it works now in the MWE!

**Comment by johannbrehmer** [comment]
With the last commit it `PyroSimulator.augmented_data()` at least *seems* to work. But I haven't tested it extensively yet.

**Comment by johannbrehmer** [comment]
I'll leave it up to you whether to merge this PR now or whether to wait for more tests.

**Review by jan-matthis — APPROVED** [review]

**Comment by jan-matthis** [comment]
Looks good to me! We should set up an explicit suite of tests for simulator classes and implementations.
